core/src/typ.rs

```diff
@@ -927,6 +927,99 @@ impl Type {
         .unwrap()
     }
 
+    /// Static typing guarantees make some of the contract checks useless, assuming that blame
+    /// safety holds. This function simplifies `self` for contract generation, assuming it is part
+    /// of a static type annotation, by eliding some of these useless subcontracts.
+    ///
+    /// # Simplifications
+    ///
+    /// - `forall`s in positive positions are removed, and the corresponding type variable is
+    ///   substituted for a `Dyn` contract. In consequence, [Self::contract()] will generate
+    ///   optimized contracts as well (for example, `forall a. Array a -> a` becomes `Array Dyn ->
+    ///   Dyn`, where `Array Dyn` will be translated to `$array_dyn` which has a constant-time
+    ///   overhead while `Array a` is linear in the size of the array.
+    /// - All positive occurrences of first order contracts (that is, anything but a function type)
+    /// are turned to `Dyn` contracts.
+    fn optimize_static(self) -> Self {
+        fn optimize_rrows(
+            rrows: RecordRows,
+            vars_elide: &HashSet<LocIdent>,
+            polarity: Polarity,
+        ) -> RecordRows {
+            RecordRows(rrows.0.map(
+                |typ| Box::new(optimize(*typ, vars_elide, polarity)),
+                |rrows| Box::new(optimize_rrows(*rrows, vars_elide, polarity)),
+            ))
+        }
+
+        fn optimize(typ: Type, vars_elide: &HashSet<LocIdent>, polarity: Polarity) -> Type {
+            let mut pos = typ.pos;
+
+            let optimized = match typ.typ {
+                TypeF::Arrow(dom, codom) => TypeF::Arrow(
+                    Box::new(optimize(*dom, vars_elide, polarity.flip())),
+                    Box::new(optimize(*codom, vars_elide, polarity)),
+                ),
+                // TODO: don't optimize only VarKind::Type
+                TypeF::Forall {
+                    var,
+                    var_kind: VarKind::Type,
+                    body,
+                } if polarity == Polarity::Positive => {
+                    let mut env = vars_elide.clone();
```

Comment:
This is technically quadratic, right? I guess we don't expect the foralls to be deeply nested, though...